### PR TITLE
fix for float truncation affecting timing in large timeframe numbers

### PR DESCRIPTION
**Pull request by noferini**
Required to simulate cosmic rays (low rates --> large number of timeframes)

### CODE REVIEW

**Comment by noferini** [comment]
@preghenella I applied the fix to run simulation of cosmic rays (already tested)

**Comment by shahor02** [comment]
@noferini instead of using potentially large abs. time, is not it better to use bc/orbit and a time within BC? Note that this is what now ``o2::InteractionTimeRecord" provides to digitizer, see how it is used in the ITS digitizer: 
https://github.com/AliceO2Group/AliceO2/blob/914de17b21e87bd591c361e07b77ac83a6792a46/Detectors/ITSMFT/common/simulation/src/Digitizer.cxx#L84
https://github.com/AliceO2Group/AliceO2/blob/914de17b21e87bd591c361e07b77ac83a6792a46/Detectors/ITSMFT/common/simulation/src/Digitizer.cxx#L185

**Comment by noferini** [comment]
@shahor02 Hi Ruben,
I am not sure to understand your comment.
In TOF digitization we receive an event_time (double) + geant_time of the hits.

Are you suggesting to split the event time in
orbit + bc + event_time_smearing
and then applying only event_time_smearing to our time?
In any case if the event_time is large the precision problem will be always limited by double precision.
Cheers
Francesco

**Comment by shahor02** [comment]
@noferini An event_time in "double" is what you construct on the flight, what "digitization context" provides for every interaction is orbit/bc and and time in ``ns`` relative to the center of the BC :
https://github.com/AliceO2Group/AliceO2/blob/914de17b21e87bd591c361e07b77ac83a6792a46/DataFormats/common/include/CommonDataFormat/InteractionRecord.h#L242-L243
Since all ALICE clocks are synced to LHC clock in one way or another, if you would use directly the InteractionRecord and this time_in_BC (augmented by the hit time wrt the collision) to map the hit to some readout window and time within this window, you are guaranteed to (1) calculate this window using int arithmetic only, w/o any precision problem, (2) to never deal with times exceeding a few 10 ns of the time of flight.

**Comment by noferini** [comment]
I see.
I didn't note that the time was converted in ns from an interactionTimeRecord in the digitizerTask.
Then I have to change what digitizer task provides to our digitizer.

Thanks for the suggestion.
Cheers
Francesco

**Comment by shahor02** [comment]
I've modified the InteractionTimeRecord ~ month ago exactly to avoid round-off errors (saw them in ITS) and announced this in wp12:
https://groups.cern.ch/group/alice-o2-wp12/Lists/Archive/Flat.aspx?RootFolder=%2fgroup%2falice%2do2%2dwp12%2fLists%2fArchive%2fchange%20of%20InteractionTimeRecord%20content&FolderCTID=0x012002009ABC3E93B6675A47B4BB3F6AFE31478F

Cheers,
Ruben

**Comment by noferini** [comment]
I switched to use InteractionTimeRecord.

Cheers
Francesco

File: Detectors/TOF/simulation/src/Digitizer.cxx [hunk 1]
```diff
@@ -271,19 +271,15 @@ void Digitizer::addDigit(Int_t channel, UInt_t istrip, Double_t time, Float_t x,
...
-  // additional check to avoid very rare truncation
-  if (tdc < 0) {
-    nbc--;
-    tdc += 1024;
-  } else if (tdc >= 1024) {
-    nbc++;
-    tdc -= 1024;
-  }
+  // add orbit and bc
+  nbc += mEventTime.orbit * Geo::BC_IN_ORBIT + mEventTime.bc;
```
Comment (shahor02) [inline Detectors/TOF/simulation/src/Digitizer.cxx:282]: just for info: you could use ``nbc += mEventTime.toLong();``

**Comment by noferini** [comment]
Nice! I'm going to apply also this suggestion.

**Comment by noferini** [comment]
Everything done.
Cheers
Francesco

**Pull request merged**